File: 1130-minimum-cost-tree-from-leaf-values/1130-minimum-cost-tree-from-leaf-values.js
/**
 * @param {number[]} arr
 * @return {number}
 */
var mctFromLeafValues = function(arr) {
  const stack = [Infinity];
  let ans = 0;
  
  for (let num of arr) {
    while (stack.at(-1) <= num) {
      const mid = stack.pop();
      ans += mid * Math.min(stack.at(-1), num);
    }
    
    stack.push(num);
  }

  for (let i = 2; i < stack.length; i++) {
    ans += stack[i] * stack[i - 1];
  }
  
  return ans;
};

File: 1130-minimum-cost-tree-from-leaf-values/1130-minimum-cost-tree-from-leaf-values.py
class Solution:
    def mctFromLeafValues(self, arr: List[int]) -> int:
        stack = [float('inf')]
        ans = 0
        for a in arr:
            while stack[-1] <= a:
                mid = stack.pop()
                ans += mid * min(stack[-1], a)
            stack.append(a)
        
        while len(stack) > 2:
            ans += stack.pop() * stack[-1]
        return ans